File: go/life/timer.go
package life

import (
	"fmt"
	"time"
)

// Timer object tracks elapsed and average times in milliseconds
type Timer struct {
	msElapsed, totElapsed float64
	avgElapsed            float64
	timerStart            time.Time
	timesCalled           int64
	prevAvg               string
	stableCount           int
}

// Start must always be called before Stop; initialises Timer
func (t *Timer) Start() {
	t.timerStart = time.Now()
}

// Stop is called once code to be timed is completed
func (t *Timer) Stop() {
	timerStop := time.Now()

	t.msElapsed = timerStop.Sub(t.timerStart).Seconds() * 1000
	t.timesCalled++
	t.totElapsed += t.msElapsed

	t.prevAvg = fmt.Sprintf("%.3f", t.avgElapsed)
	t.avgElapsed = t.totElapsed / float64(t.timesCalled)
}

// Elapsed returns elapsed milliseconds between Start() and Stop()
func (t *Timer) Elapsed() string {
	return fmt.Sprintf("%.3f", t.msElapsed)
}

// AverageElapsed returns average duration over all timings
func (t *Timer) AverageElapsed() string {
	return fmt.Sprintf("%.3f", t.avgElapsed)
}

// TotalElapsed returns total duration of all timings
func (t *Timer) TotalElapsed() string {
	return fmt.Sprintf("%.3f", t.totElapsed)
}

// String provides automatic stringification for Timer
func (t *Timer) String() string {
	return fmt.Sprintf("Loop %sms (Avg %sms)", t.Elapsed(), t.AverageElapsed())
}

// IsAverageStable returns true when AverageElapsed() remains constant n times
func (t *Timer) IsAverageStable(n int) bool {
	currAvg := fmt.Sprintf("%.3f", t.avgElapsed)
	if currAvg == t.prevAvg {
		t.stableCount++
	} else {
		t.stableCount = 0
	}
	return t.stableCount >= n
}

File: python/life/timer.py
import time

class Timer:
    msElapsed = 0
    totElapsed = 0
    avgElapsed = 0
    timerStart = 0
    timesCalled = 0
    prevAvg = ""
    stableCount = 0

    def start(self):
        self.timerStart = time.time_ns() / 1000000.0

    def stop(self):
        timerStop = time.time_ns() / 1000000.0
        if self.timerStart != 0:
            self.msElapsed = timerStop - self.timerStart
            self.timesCalled += 1
            self.totElapsed += self.msElapsed

            self.prevAvg = "{:.3f}".format(self.avgElapsed)
            self.avgElapsed = self.totElapsed / self.timesCalled

    def elapsed(self):
        return "{:.3f}".format(self.msElapsed)

    def averageElapsed(self):
        return "{:.3f}".format(self.avgElapsed)

    def totalElapsed(self):
        return "{:.3f}".format(self.totElapsed)

    def isAverageStable(self, n):
        currAvg = self.averageElapsed()
        if currAvg == self.prevAvg:
            self.stableCount += 1
        else:
            self.stableCount = 0
        return self.stableCount >= n

    def __str__(self):
        return "Loop " + self.elapsed() + "(Avg " + self.averageElapsed() + ")"
